### PR TITLE
Design improvements

### PR DESCRIPTION
**Pull request by janno42**
This makes some changes to the UI. Especially the student voting page and the email sending form have changed.

**Old email form:**
email_old

**New email form:**
email_new

**Old voting page:**
vote_old

**New voting page:**
vote_new

### CODE REVIEW

File: evap/evaluation/templatetags/morefilters.py [hunk 1]
```diff
@@ -70,3 +70,19 @@ def render(self, context):
...
+def is_multiple_choice_field(field):
+    return field.field.__class__.__name__ == "MultipleChoiceField"
+
+@register.filter
+def is_textfield(field):
+    return field.field.widget.__class__.__name__ == "TextInput"
+
+@register.filter
+def is_textarea(field):
+    return field.field.widget.__class__.__name__ == "Textarea"
```
Comment (karyon) [inline evap/evaluation/templatetags/morefilters.py]: are you sure you need these? can't you check for the fields name or something like that? like, "if this field is the body field, then make a custom input", not "if this field is a textarea, make a custom input"

File: evap/staff/forms.py [hunk 2]
```diff
@@ -120,15 +117,9 @@ def __init__(self, *args, **kwargs):
...
+        self.recipient_groups = self.cleaned_data.get('recipients')
 
-        if self.cleaned_data.get('sendToAllParticipants'): self.recipient_groups += ['all_participants']
-        if self.cleaned_data.get('sendToDueParticipants'): self.recipient_groups += ['due_participants']
-        if self.cleaned_data.get('sendToResponsible'): self.recipient_groups += ['responsible']
-        if self.cleaned_data.get('sendToEditors'): self.recipient_groups += ['editors']
-        if self.cleaned_data.get('sendToContributors'): self.recipient_groups += ['contributors']
-
-        if len(self.recipient_groups) == 0:
+        if not self.recipient_groups:
```
Comment (karyon) [inline evap/staff/forms.py]: like.

do you know whether we have the option to put custom text into the buttons? right now the text comes from EMAIL_RECIPIENTS, right?

File: evap/staff/templates/staff_course_email.html [hunk 3]
```diff
@@ -2,18 +2,52 @@
 
 {% load i18n %}
 {% load static %}
+{% load morefilters %}
 
 {% block content %}
     {{ block.super }}
+    {% if form.errors %}
+        <div class="alert alert-danger" role="alert">{% blocktrans %}Please fill out all required fields and select at least one recipient group.{% endblocktrans %}</div>
+    {% endif %}
```
Comment (karyon) [inline evap/staff/templates/staff_course_email.html]: is this necessary? i mean the form is small enough and the errors of the individual fields should be explicit enough, or aren't they?

File: evap/staff/forms.py (hunk 2 above)
Comment (karyon) [inline evap/staff/forms.py]: oh wait... forget what i said.

File: evap/student/templates/student_vote.html [hunk 4]
```diff
@@ -14,8 +14,8 @@
 
 {% block content %}
     {{ block.super }}
-    {% if errors %}
-        <div class="alert alert-danger" role="alert">{% blocktrans %}Please make sure to vote for all contributors or click on "I can't give feedback" to skip a person.{% endblocktrans %}</div>
+    {% if is_bound %}
```
Comment (karyon) [inline evap/student/templates/student_vote.html]: why is_bound? aren't there bound forms without errors?

Comment (karyon) [inline evap/student/templates/student_vote.html]: ok, i now know why it works, but why didn't "if errors" work as well? seems more explicit to me.

File: evap/student/templates/student_vote.html [hunk 5]
```diff
@@ -82,10 +79,7 @@ <h3 class="panel-title">{{ contributor.full_name }}</h3>
                                         <table class="table table-striped">
                                             <tbody>
                                                 {% for field in form %}
-                                                    <tr class="vote-row">
-                                                        <td>{{ field.label }}</td>
-                                                        <td class="vote-inputs {% if preview %}preview{% endif %}">{{ field.errors }}{{ field }}</td>
-                                                    </tr>
+                                                    {% include "student_vote_row.html" %}
```
Comment (karyon) [inline evap/student/templates/student_vote.html]: wow

such amaze

Comment (karyon) [inline evap/student/templates/student_vote.html]: you need to create a inclusion tag for that. including templates without such a tag (and without the template cache) costs a lot of performance if doing it multiple times on one page.

File: evap/student/templates/student_vote_row.html [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+                <label class="btn btn-sm btn-default vote-btn {% if field.value == choice.choice_value %}active{% endif %} {% if field.errors %}choice-error{% endif %}" name="{{ choice.name }}" onclick="selectedAnswer('{{ choice.name }}');">
+                    <input id="{{ choice.id_for_label }}" name="{{ choice.name }}" type="radio" value="{{ choice.choice_value }}" autocomplete="off" {% if field.value == choice.choice_value %}checked="checked"{% endif %}/>
+                    {{ choice.choice_label }} {{ choice.id }}
+                </label>
+            {% endfor %}
+        </td>
+    {% else %}
+        <td class="vote-inputs {% if preview %}preview{% endif %}">{{ field.errors }}{{ field }}</td>
+    {% endif %}
+</tr>
```
Comment (karyon) [inline evap/student/templates/student_vote_row.html]: ooohh what's that

File: evap/staff/templates/staff_course_email.html (hunk 3 above)
Comment (janno42) [inline evap/staff/templates/staff_course_email.html]: error texts are not shown anymore. the respective fields are just marked in red. so i felt some message is needed.

File: evap/student/templates/student_vote.html (hunk 4 above)
Comment (janno42) [inline evap/student/templates/student_vote.html]: errors just holds the contributor errors

File: evap/student/templates/student_vote_row.html (hunk 6 above)
Comment (janno42) [inline evap/student/templates/student_vote_row.html]: what exactly do you mean? ;)

Comment (karyon) [inline evap/student/templates/student_vote_row.html]: there is no newline at end of file.

File: evap/student/templates/student_vote.html (hunk 4 above)
Comment (karyon) [inline evap/student/templates/student_vote.html]: alright then please rename that variable to contributor_errors.

still, i'd like something more explicit than is_bound here. maybe look whether there are any errors in the general contribution, and then do "if general_contribution_errors or contributor_errors" here or something like this.

Comment (karyon) [inline evap/student/templates/student_vote.html]: regarding the other template parameters, course_forms and contributor_questionnaires sound like different things but the former is just the general contribution, right? could you check that and then call them [course/contributor]_forms? or maybe [course/contributor]_questionnaires, not sure which one is more accurate.

**Comment by karyon** [comment]
the contributor's course preview view is not accessible right now, but still exists. please fix the template parameters there as well.

File: evap/evaluation/templatetags/morefilters.py [hunk 7]
```diff
@@ -70,3 +70,7 @@ def render(self, context):
 @register.filter
 def is_false(arg): 
     return arg is False
+
+@register.filter
+def is_choice_field(field):
+    return field.field.__class__.__name__ == "TypedChoiceField"
```
Comment (karyon) [inline evap/evaluation/templatetags/morefilters.py]: did you not remove this intentionally?

Comment (janno42) [inline evap/evaluation/templatetags/morefilters.py]: here it's a different situation than in the other case. from the ids and names of the questions i can't see whether they need a textfield or a choice field. so i have to look that up somehow else.

File: evap/student/templates/student_index.html [hunk 8]
```diff
@@ -72,7 +72,7 @@ <h3 class="panel-title">{{ semester.semester_name }}</h3>
                                 </td>
                                 <td>
                                     {% if course.state == 'inEvaluation' and course in due_courses %}
-                                        {% if course.days_left_for_evaluation == 0 %}
+                                        {% if course.days_left_for_evaluation <= 0 %}
```
Comment (karyon) [inline evap/student/templates/student_index.html]: uh what?

**Comment by karyon** [comment]
i'm done, if you could review the last three commits and make sure the collapse-things works as you expect?

**Comment by janno42** [comment]
the collapsing is nice, but the preview for contributors and staff is empty. it holds the contributor names but no questions at all. could you please check why?

**Comment by karyon** [comment]
thank you, fixed.

**Comment by janno42** [comment]
please remove the one line and i think then we're done :+1:

**Pull request merged**